RetailCoder.VBE/UI/Command/Refactorings/FormDesignerRefactorRenameCommand.cs

```diff
@@ -45,35 +43,38 @@ protected override void OnExecute(object parameter)
         {
-            var project = _vbe.ActiveVBProject;
-            var component = _vbe.SelectedVBComponent;
+            var projectId = qualifiedModuleName.HasValue
+                                               ? qualifiedModuleName.Value.ProjectId
+                                               : Vbe.ActiveVBProject.ProjectId;
+            
+            var component = qualifiedModuleName.HasValue
+                                               ? qualifiedModuleName.Value.Component
+                                               : Vbe.SelectedVBComponent;
+                        
+            if (component?.HasDesigner ?? false)
             {
-                if (Vbe.SelectedVBComponent != null && Vbe.SelectedVBComponent.HasDesigner)
+                if (qualifiedModuleName.HasValue)
                 {
-                    var designer = ((dynamic)component.Target).Designer;
-
-                    if (designer.selected.count == 1)
-                    {
-                        var control = designer.selected.item(0);
-                        var result = _state.AllUserDeclarations
-                            .FirstOrDefault(item => item.DeclarationType == DeclarationType.Control
-                                                    && project.HelpFile == item.ProjectId
-                                                    && item.ComponentName == component.Name
-                                                    && item.IdentifierName == control.Name);
-
-                        Marshal.ReleaseComObject(control);
-                        Marshal.ReleaseComObject(designer);
-                        return result;
-                    } else {
-                        var message = string.Format(RubberduckUI.RenameDialog_AmbiguousSelection);
-                        _messageBox.Show(message, RubberduckUI.RenameDialog_Caption, MessageBoxButtons.OK,
-                            MessageBoxIcon.Exclamation);
-                    }
+                    return _state.DeclarationFinder.MatchName(qualifiedModuleName.Value.Name)
+                                                   .SingleOrDefault(m => m.ProjectId == projectId
+                                                                      && m.DeclarationType.HasFlag(qualifiedModuleName.Value.ComponentType)
+                                                                      && m.ComponentName == component.Name);
                 }
-            }
-
+                
+                var selectedCount = component.SelectedControls.Count;
+                if (selectedCount > 1) { return null; }
+                
+                // Cannot use DeclarationType.UserForm, parser only assigns UserForms the ClassModule flag
+                var selectedType = selectedCount == 0 ? DeclarationType.ClassModule : DeclarationType.Control;
+                var selectedName = selectedCount == 0 ? component.Name : component.SelectedControls[0].Name;
```

Comment:
I think I'd find this cleaner with an if-statement instead of two separate ternaries with the same condition... After all the same condition usually means the values are related...

Reply:
At that point... can we use tuple destructuring here?

    selectedType, selectedName= selectedCount == 0 ? (DeclarationType.ClassModule, component.Name) : (DeclarationType.Control, component.SelectedControls[0].Name);